substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/service/ServiceThread.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) 2024, 2024, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2024, 2024, Red Hat Inc. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.oracle.svm.core.genscavenge.service;
+
+import com.oracle.svm.core.Uninterruptible;
+import com.oracle.svm.core.jdk.UninterruptibleUtils;
+import com.oracle.svm.core.locks.VMSemaphore;
+import com.oracle.svm.core.log.Log;
+import org.graalvm.nativeimage.Platform;
+import org.graalvm.nativeimage.Platforms;
+
+/**
+ * This class is the dedicated thread that handles services. For now, the only service is GC
+ * notifications.
+ */
+public class ServiceThread extends Thread {
+    private final UninterruptibleUtils.AtomicBoolean atomicNotify;
+    private final VMSemaphore semaphore;
```

Comment:
Please use `VMMutex` and `VMCondition` instead (we only use `VMSemaphore` where it is absolutely necessary to do so, as it is much more complex to create on some platforms).

Reply:
I replaced the usage of  `VMSemaphore` with `VMMutex` and `VMCondition`. However, I experienced deadlock in `NotificationThread.signal()` when trying to lock the mutex during GC. So instead I'm calling `VMCondition.signal()` without holding the mutex. So there can be a rare (and probably harmless) race described in [this comment](https://github.com/oracle/graal/pull/9799/files#diff-2849899c2604e9a87436de2bf07ffbabb5c0d878c9de12eafd33d435b7eea879R79-R85).

Do you think there's a better way to do this?